bionic/core/flow_execution.py

```diff
@@ -108,32 +92,59 @@ def run(self, states):
             if self._parallel_execution_enabled:
                 self._bootstrap.executor.stop_logging()
 
-    def _process_entry(self, entry):
+    def _process_active_entry(self, entry):
+        """
+        Takes the current active entry and does the minimal amount of work to move it
+        to the next stage. When this function returns, the entry will no longer be in
+        the ACTIVE stage: it will be COMPLETED, IN_PROGRESS, or BLOCKED, and we will
+        be able to move on to the next entry.
+        """
+
         assert entry.stage == EntryStage.ACTIVE
 
-        state = entry.state
-        # Initialize the task state before attempting to complete it.
-        state.initialize(self._bootstrap, self._flow_instance_uuid)
+        # In theory we could have an entry that doesn't require priming, but in practice
+        # we always require it when we create one.
+        assert entry.is_requested_to_be_primed
 
-        # Attempt to complete the task state from persistence cache.
-        state.attempt_to_complete_from_cache()
-        if state.is_cached:
+        # First, if we already have a cached value, we can stop immediately without
+        # needing to process any dependencies.
+        entry.state.set_up_caching_flags(self._bootstrap)
+        if entry.is_cached:
             self._mark_entry_completed(entry)
-
-        elif (
-            not self._parallel_execution_enabled
-            or entry.is_needed_in_memory
-            # This is a simple lookup task that looks up a value in a dictionary.
-            # We can't run this in a separate process because the value may not be
-            # cloudpicklable.
-            or state.task.is_simple_lookup
-        ):
-            entry.compute(self.task_key_logger)
-            self._mark_entry_completed(entry)
-
-        # Compute the results for serializable entity in parallel.
-        elif state.should_persist:
-            new_state_for_subprocess = state.strip_state_for_subprocess()
+            return
+
+        # Otherwise, we'll need all of our dependencies to be primed so that their
+        # digests are available.
+        self._set_up_entry_dependencies(entry)
+        for dep_entry in entry.dep_entries:
+            self._request_entry_be_primed(dep_entry)
+        if not entry.all_dependencies_satisfy_requests:
+            self._mark_entry_blocked(entry)
```

Comment:
`_mark_entry_blocked` also tries to mark blocking entries pending, which `_request_entry_be_primed` and `_request_entry_be_cached` already do it for you. Can we combine those logic as well, maybe assert that blocking deps are not completed instead of marking them as pending?

Reply:
Yes, good point -- thanks!